core/src/main/java/org/apache/iceberg/SnapshotParser.java

```diff
@@ -106,6 +124,10 @@ public static String toJson(Snapshot snapshot, boolean pretty) {
   }
 
   static Snapshot fromJson(JsonNode node) {
+    return fromJson(node, null);
+  }
+
+  static Snapshot fromJson(JsonNode node, EncryptionManager encryption) {
```

Comment:
I don't think that this should use the encryption manager. `SnapshotParser` should only be responsible for encoding and decoding the snapshot fields, not for encrypting and decrypting. The logic added here is really complex and should be handled elsewhere.